design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+Consuming projections should also retrieve the error info associated with the xlang error and associate it in some way with the native error.  It should be associated in a way that allows the logging of the native error to also log the information in the error info.  For instance, in certain languages, exceptions have a ToString function which consumers of that language call upon failure to log the error for later analysis.  This ToString function should also return the information in xlang's error info to allow the developer to get a better picture of the failure.  In addition, it should also be associated in a way that allows a producing projection for the same language to determine if there is an error info associated with the native error and to be able to retrieve it if there is.  This will allow the producing projection to propagate the error with the previously collected diagnostic information if the error ends up being propagated by the component instead of being handled.
+
+It is left to the projections to decide how to achieve this because it will be different for each language.  But this design note will suggest some approaches to consider.  One approach is using inheritance in the case of exception-based projections if the language isn't affected by object slicing.  For each of the errors from xlang that the projection maps to, a new type inheriting the native exception type for that error can be created and be used to store the error info.  This new type would just be an implementation detail of the projection and would not be caught directly by components written in that language.  Instead components would use the native exception types which these types inherit from to catch exceptions.  Projections would use these custom types to determine whether it is a translated error and to get the associated error info.  For error code based projections, an approach would be to use a thread level storage to store the error info.  The consuming projection would store the error info there, while the producing projection would retrieve it from there when the same error propagates to the producing side.  If the error is handled in the consumer, the consumer should have an API provided by the projection to clear the stored error info to ensure proper cleanup of it.  The consumer should also have an API available to get/set the stored error info if the error gets passed across threads or to log it if a failure occurs.
```

Comment:
Consider breaking each of these approaches into its own \#\#\# section.

Reply:
Addressed.